router: fix upstream trace spans not containing gRPC annotations

nareddyt opened the pull request:
Commit Message:
router: fix upstream trace spans not containing gRPC annotations

Additional Description:

When `UpstreamRequest` is cleaned up, the upstream headers/trailers are used to add gRPC-related trace annotations to the upstream span.

https://github.com/envoyproxy/envoy/blob/ce49c7f65668a22b80d1e83c35d170741bb8d46a/source/common/router/upstream_request.cc#L85

But the upstream headers/trailers were never initialized. This PR fixes the bug to set upstream gRPC tags as soon as headers/trailers are available.

Now upstream gRPC spans have two extra annotations (as intended):

- `grpc.message`
- `grpc.status_code`

Risk Level: Low, just adds new fields to spans

Testing: Unit test

Docs Changes: None, most users probably never noticed the bug

Release Notes: None, no new features

Platform Specific Features: None

Signed-off-by: nareddyt <nareddyt@google.com>

nareddyt commented:
cc @qiwzhang @TAOXUY @shuoyang2016 @dafudeng

nareddyt commented:
Please take a look @wbpcode

nareddyt commented:
Please take a look @wbpcode

wbpcode commented:
Sorry for the delay. I will review it tomorrow morning.

nareddyt commented on test/common/router/upstream_request_test.cc:
```diff
@@ -3,37 +3,46 @@
+  NiceMock<MockTimeSystem> time_system;
+  NiceMock<Server::Configuration::MockFactoryContext> context_;
   NiceMock<MockRouterFilterInterface> router_filter_interface_;
-  UpstreamRequest upstream_request_{router_filter_interface_,
-                                    std::make_unique<NiceMock<Router::MockGenericConnPool>>(),
-                                    false, true};
```
For the `DecodeHeadersGrpcSpanAnnotations`, I need to modify the route config to set `start_child_span`. This option is cached in `UpstreamRequest` as soon as the constructor runs. So I prevent the constructor from running too early by moving the construction of `UpstreamRequest` to the test itself. This gives the test a chance to modify the config before construction.

wbpcode replied:
Then it would be better to add a simple helper function to do this to reduce there repeated code snippets.

nareddyt replied:
Done, this is now part of `initialize()` after merge with previous commits.

nareddyt commented:
PTAL @wbpcode

wbpcode commented:
This is lgtm overall for me. Only one comment is added.

wbpcode commented:
cc @alyssawilk Could you take a second review for this core code update when you have free time?

alyssawilk commented:
Thanks for tackling this!  Your implementation looks correct, but it's a shame that we have to make a full copy of the headers and trailers to extract 2 tags for tracing.  
What if we removed headers/trailers from finalize and added 2 optional interim utilities
onResponseHeaders
onResponseTrailers
which we could call when the routre received these two. Then we could do the tag extraction early, and finalize the span in clean up, without worrying about doing a full copy of both headers and trailers.
/wait

nareddyt commented:
> Thanks for tackling this! Your implementation looks correct, but it's a shame that we have to make a full copy of the headers and trailers to extract 2 tags for tracing. What if we removed headers/trailers from finalize and added 2 optional interim utilities onResponseHeaders onResponseTrailers which we could call when the routre received these two. Then we could do the tag extraction early, and finalize the span in clean up, without worrying about doing a full copy of both headers and trailers. /wait

Apologies for reviving this so late.

This is a great idea, done. PTAL

alyssawilk commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -210,7 +210,9 @@ void HttpTracerUtility::finalizeDownstreamSpan(Span& span,
 
-  setCommonTags(span, response_headers, response_trailers, stream_info, tracing_config);
+  setCommonTags(span, stream_info, tracing_config);
+  onUpstreamResponseHeaders(span, response_headers);
```
should these still be called here?  is there a chance that gRPC tags will be added twice or is that safe?

nareddyt replied:
> should these still be called here?

Yes. Envoy supports gRPC downstream and currently sets the gRPC tags for the "main" downstream span. This preserves the behavior.

> is there a chance that gRPC tags will be added twice or is that safe?

Should be safe. gRPC spec guarantees `grpc-status` only occurs once. https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md

alyssawilk replied:
Awesome, that's what I needed to know, thanks!

alyssawilk reviewed (commented):
Cool, I think this is much cleaner!
/wait-any

alyssawilk commented:
@lizan ping?

alyssawilk commented:
oops, @snowp could you do a quick pass instead?

alyssawilk commented:
@snowp ping!

snowp reviewed (approved):
Makes sense to me, thanks!
